cmd/autoretrieve/filecoin/retriever.go

```diff
@@ -0,0 +1,393 @@
+package filecoin
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"path/filepath"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/application-research/estuary/cmd/autoretrieve/blocks"
+	"github.com/application-research/filclient"
+	"github.com/application-research/filclient/keystore"
+	"github.com/application-research/filclient/retrievehelper"
+	"github.com/dustin/go-humanize"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-fil-markets/retrievalmarket"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/wallet"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	"golang.org/x/xerrors"
+)
+
+var logger = log.Logger("autoretrieve")
+
+const walletSubdir = "wallet"
+
+var (
+	ErrNoCandidates            = errors.New("no candidates")
+	ErrRetrievalAlreadyRunning = errors.New("retrieval already running")
+)
+
+type RetrieverConfig struct {
+	DataDir          string
+	Endpoint         string
+	MinerBlacklist   map[address.Address]bool
+	RetrievalTimeout time.Duration
+}
+
+type Retriever struct {
+	config              RetrieverConfig
+	filClient           *filclient.FilClient
+	runningRetrievals   map[cid.Cid]bool
+	runningRetrievalsLk sync.Mutex
+}
+
+type retrievalCandidate struct {
+	Miner   address.Address
+	RootCid cid.Cid
+	DealID  uint
+}
+
+type candidateQuery struct {
+	candidate retrievalCandidate
+	response  *retrievalmarket.QueryResponse
+}
+
+func NewRetriever(config RetrieverConfig, host host.Host, api api.Gateway, datastore datastore.Batching, blockManager *blocks.Manager) (*Retriever, error) {
+
+	keystore, err := keystore.OpenOrInitKeystore(filepath.Join(config.DataDir, walletSubdir))
+	if err != nil {
+		return nil, err
+	}
+
+	wallet, err := wallet.NewWallet(keystore)
```

Comment:
in general, instead of instantiating things inside constructors, pass them into the constructor. In this case, the wallet and filclient instance should just be taken as parameters to the NewRetriever function

Reply:
hmm, interesting thing to consider. i made it this way because i wanted to keep things that didn't need to be shared inside of their own modules to avoid spaghetti. it doesn't matter if i expose them at the top level since this isn't a library. 

i do see how this might not be future proof, though, in the case that i might want to add another module which does need filclient and wallet. are there any other important reasons?